design-patterns/behavioral-patterns/src/iterator/IcecreamRepository.java

```diff
@@ -0,0 +1,52 @@
+package iterator;
+
+import java.util.Iterator;
+
+public class IcecreamRepository implements Iterable<String> {
+
+	private String [] icecreams;
+	private int index;
+	
+	public IcecreamRepository() {
+		icecreams = new String[10];
+		index = 0;
+	}
+	
+	public void addIcecream(String newIcecream) {
+		if(index == icecreams.length) {
+			String [] moreIcecreams = new String[icecreams.length + 5];
+			System.arraycopy( icecreams, 0, moreIcecreams, 0, icecreams.length );
+			icecreams = moreIcecreams;
+			moreIcecreams = null;
+		}
+		
+		icecreams[index++] = newIcecream;
+	}
+
+	@Override
+	public Iterator<String> iterator() {
+		Iterator<String> it = new Iterator<String>() {
+
+			private int currentIndex = 0;
+			
+			@Override
+			public boolean hasNext() {
+				return currentIndex < icecreams.length && icecreams[currentIndex] != null;
```

Comment:
You will have a problem in the implementation. if input some ice creams and one of them is 'null' the iteration will not work properly. For Example:
```
new String["vanila", null, "chocolate"]
```

With this array of ice creams the iteration will stop at the second one and the only the first one will be rendered.
You have two options here to not allow null elements in the array or to allow iteration of the null elements

Reply:
The check for null is no longer necessary as we are sure that the array does not contains null elements